metricproviders/datadog/datadog.go

```diff
@@ -0,0 +1,172 @@
+	return ProviderType
+}
+
+// Run queries DataDog for a metric
+func (p *Provider) Run(run *v1alpha1.AnalysisRun, metric v1alpha1.Metric) v1alpha1.Measurement {
+	startTime := metav1.Now()
+	newMeasurement := v1alpha1.Measurement{
+		StartedAt: &startTime,
+	}
+
+	query, err := templateutil.ResolveArgs(metric.Provider.Datadog.Query, run.Spec.Args)
+	if err != nil {
+		return metricutil.MarkMeasurementError(newMeasurement, err)
+	}
+
+	// TODO (cabrinha) make from and to configurable
+	from := time.Now().Unix() - 60
+	to := time.Now().Unix()
+	response, err := p.client.QueryMetrics(from, to, query)
+	if err != nil {
+		return metricutil.MarkMeasurementError(newMeasurement, err)
+	}
+
+	// If datadog returns an empty []Series, retry with a longer window
+	var retries int
+	for retries = 0; retries < 3 && len(response) < 1; retries++ {
+		response, err = p.client.QueryMetrics(from-60, to, query)
+		if err != nil {
+			return metricutil.MarkMeasurementError(newMeasurement, err)
+		}
+
+		if len(response) < 1 {
+			time.Sleep(1 * time.Second)
+		}
+	}
+
+	if len(response) > 0 {
+		newValue, newStatus, err := p.processResponse(metric, response[0])
+		if err != nil {
+			return metricutil.MarkMeasurementError(newMeasurement, err)
+		}
+		newMeasurement.Value = newValue
+		newMeasurement.Phase = newStatus
+		finishedTime := metav1.Now()
+		newMeasurement.FinishedAt = &finishedTime
+		return newMeasurement
+	}
+	newMeasurement.Value = ""
+	newMeasurement.Phase = v1alpha1.AnalysisPhaseInconclusive
+	finishedTime := metav1.Now()
+	newMeasurement.FinishedAt = &finishedTime
+	return newMeasurement
+}
+
+// Resume should not be used with the datadog provider since all the work should occur in the Run method
+func (p *Provider) Resume(run *v1alpha1.AnalysisRun, metric v1alpha1.Metric, measurement v1alpha1.Measurement) v1alpha1.Measurement {
+	p.logCtx.Warn("Datadog provider should not execute the Resume method")
+	return measurement
+}
+
+// Terminate should not be used with the datadog provider since all the work should occur in the Run method
+func (p *Provider) Terminate(run *v1alpha1.AnalysisRun, metric v1alpha1.Metric, measurement v1alpha1.Measurement) v1alpha1.Measurement {
+	p.logCtx.Warn("Datadog provider should not execute the Terminate method")
+	return measurement
+}
+
+// GarbageCollect is a no-op for the datadog provider
+func (p *Provider) GarbageCollect(run *v1alpha1.AnalysisRun, metric v1alpha1.Metric, limit int) error {
+	return nil
+}
+
+func (p *Provider) processResponse(metric v1alpha1.Metric, response dd.Series) (string, v1alpha1.AnalysisPhase, error) {
+	length := len(response.Points)
+	switch {
+	case length < 1:
+		return "", v1alpha1.AnalysisPhaseInconclusive, nil
+	// If we get 1 or more points, select the first point
+	case length >= 1:
+		result := *response.Points[0][1]
+		valueStr := fmt.Sprintf("%f", result)
+		newStatus := evaluate.EvaluateResult(result, metric, p.logCtx)
+		return valueStr, newStatus, nil
```

Comment:
Since the `evaluate.EvaluateResult` can evaluate expressions like `all(result, {#>1}`, we should pass the unfiltered result

Reply:
Example of an array evaluation: https://github.com/argoproj/argo-rollouts/blob/master/utils/evaluate/evaluate_test.go#L119-L124.

If a user wants to evaluate the first result only, they can write a query like `result[0] > .99`

Reply:
Alright, I've created the flattenResponse() func that just takes the `[]DataPoint`s and then flattens it to a `[]float64` that only contains the actual data point values.